pkg/storagev2/internal/upload/uploadstore.go

```diff
@@ -0,0 +1,166 @@
+// Copyright 2022 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package upload
+
+import (
+	"context"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/ethersphere/bee/pkg/storagev2"
+	"github.com/ethersphere/bee/pkg/storagev2/internal"
+	"github.com/ethersphere/bee/pkg/swarm"
+)
+
+// now returns the current time.Time; used in testing.
+var now = time.Now
+
+var (
+	// errTagIDAddressItemMarshalAddressIsZero is returned when trying
+	// to marshal a tagIDAddressItem with an address that is zero.
+	errTagIDAddressItemMarshalAddressIsZero = errors.New("marshal tagIDAddressItem: address is zero")
+
+	// errTagIDAddressItemUnmarshalInvalidSize is returned when trying
+	// to unmarshal buffer that is not of size tagIDAddressItemSize.
+	errTagIDAddressItemUnmarshalInvalidSize = errors.New("unmarshal tagIDAddressItem: invalid size")
+)
+
+// tagIDAddressItemSize is the size of a marshaled tagIDAddressItem.
+const tagIDAddressItemSize = swarm.HashSize + 8
+
+var _ storage.Item = (*tagIDAddressItem)(nil)
+
+// tagIDAddressItem is an store.Item that stores addresses of already seen chunks.
+type tagIDAddressItem struct {
+	TagID   uint64
+	Address swarm.Address
+}
+
+// ID implements the storage.Item interface.
+func (i tagIDAddressItem) ID() string {
+	return fmt.Sprintf("%d_%s", i.TagID, i.Address.ByteString())
```

Comment:
The TagID can be moved to the Namespace function. This way, we don't need to have this in the value. Check the chunkStampItem in the chunkstore PR. As the tag is the same for all the chunks in this tag, it can be part of the prefix as we would know this before we query the item.

This way we dont need to store the duplicate tagIDs in the store for each chunk. Will save some space.